File: gateware/src/tiliqua/flash.py
#!/usr/bin/env python3

"""
Flash tool for Tiliqua bitstream archives.
"""

import argparse
import json
import os
import subprocess
import sys
import tarfile

# Flash memory map constants
BOOTLOADER_BITSTREAM_ADDR = 0x000000
SLOT_BITSTREAM_BASE      = 0x100000  # First user slot starts here
SLOT_SIZE                = 0x100000
MANIFEST_SIZE            = 1024
FIRMWARE_BASE_SLOT0      = 0x1C0000
MAX_SLOTS                = 8
FLASH_PAGE_SIZE          = 1024 # 1KB

def flash_file(file_path, offset, file_type="auto", dry_run=True):
    """Flash a file to the specified offset using openFPGALoader."""
    cmd = [
        "sudo", "openFPGALoader", "-c", "dirtyJtag",
        "-f", "-o", f"{hex(offset)}",
    ]
    if file_type != "auto":
        cmd.extend(["--file-type", file_type])
    cmd.append(file_path)
    
    if dry_run:
        return cmd
    else:
        print(f"Flashing to {hex(offset)}:")
        print("\t$", " ".join(cmd))
        subprocess.check_call(cmd)

def check_region_overlaps(regions, slot=None):
    """
    Check for overlapping regions in flash commands and slot boundaries.
    Each region is aligned up before checking.
    Returns (bool, str) tuple: (has_overlap, error_message)
    """
    # Convert regions to (start, end) tuples, with aligned sizes
    aligned_regions = []
    for r in regions:
        start = r['addr']
        size = (r['size'] + FLASH_PAGE_SIZE - 1) & ~(FLASH_PAGE_SIZE - 1)  # Align up
        end = start + size
        aligned_regions.append((start, end, r['name']))

        # For non-XIP firmware, check if any region exceeds its slot
        if slot is not None:
            slot_start = (start // SLOT_SIZE) * SLOT_SIZE
            slot_end = slot_start + SLOT_SIZE
            if end > slot_end:
                return (True, f"Region '{r['name']}' exceeds slot boundary: "
                            f"ends at 0x{end:x}, slot ends at 0x{slot_end:x}")

    # Sort by start address
    aligned_regions.sort()

    # Check for overlaps
    for i in range(len(aligned_regions) - 1):
        curr_end = aligned_regions[i][1]
        next_start = aligned_regions[i + 1][0]
        if curr_end > next_start:
            return (True, f"Overlap detected between '{aligned_regions[i][2]}' (ends at 0x{curr_end:x}) "
                         f"and '{aligned_regions[i+1][2]}' (starts at 0x{next_start:x})")
    return (False, "")

def flash_archive(archive_path, slot=None, noconfirm=False):
    """
    Flash a bitstream archive to the specified slot.
    For XIP firmware, slot must be None as it can only go in the bootloader slot (0x0).
    """
    commands_to_run = []
    regions_to_check = []

    # Extract archive to temporary location
    with tarfile.open(archive_path, "r:gz") as tar:
        # Read manifest first
        manifest_info = tar.getmember("manifest.json")
        manifest_f = tar.extractfile(manifest_info)
        manifest = json.load(manifest_f)

        # Check if this is an XIP firmware
        has_xip_firmware = False
        if manifest.get("regions"):
            for region in manifest["regions"]:
                if region.get("spiflash_src") is not None:
                    has_xip_firmware = True
                    xip_offset = region["spiflash_src"]
                    break

        if has_xip_firmware:
            if slot is not None:
                print("Error: XIP firmware bitstreams must be flashed to bootloader slot")
                print(f"Remove --slot argument to flash at 0x0 with firmware at 0x{xip_offset:x}")
                sys.exit(1)
        else:
            if slot is None:
                print("Error: Must specify slot for non-XIP firmware bitstreams")
                sys.exit(1)

        # Create temp directory for extracted files
        import tempfile
        with tempfile.TemporaryDirectory() as tmpdir:
            tar.extractall(tmpdir)

            if has_xip_firmware:
                print("\nPreparing to flash XIP firmware bitstream to bootloader slot...")
                # Collect commands for bootloader location
                commands_to_run.append(
                    flash_file(
                        os.path.join(tmpdir, "top.bit"),
                        BOOTLOADER_BITSTREAM_ADDR,
                        dry_run=True
                    )
                )
                # Add bootloader bitstream region
                regions_to_check.append({
                    'addr': BOOTLOADER_BITSTREAM_ADDR,
                    'size': os.path.getsize(os.path.join(tmpdir, "top.bit")),
                    'name': 'bootloader bitstream'
                })

                # Collect commands for XIP firmware
                # Add XIP firmware regions
                for region in manifest["regions"]:
                    if "filename" not in region:
                        continue
                    commands_to_run.append(
                        flash_file(
                            os.path.join(tmpdir, region["filename"]),
                            region["spiflash_src"],
                            "raw",
                            dry_run=True
                        )
                    )
                    regions_to_check.append({
                        'addr': region["spiflash_src"],
                        'size': region["size"],
                        'name': f"firmware '{region['filename']}'"
                    })
            else:
                print(f"\nPreparing to flash bitstream to slot {slot}...")
                # Calculate addresses for this slot
                slot_base = SLOT_BITSTREAM_BASE + (slot * SLOT_SIZE)
                bitstream_addr = slot_base
                manifest_addr = (slot_base + SLOT_SIZE) - MANIFEST_SIZE
                firmware_base = FIRMWARE_BASE_SLOT0 + (slot * SLOT_SIZE)

                # Add bitstream region
                regions_to_check.append({
                    'addr': bitstream_addr,
                    'size': os.path.getsize(os.path.join(tmpdir, "top.bit")),
                    'name': 'bitstream'
                })

                # Add manifest region
                regions_to_check.append({
                    'addr': manifest_addr,
                    'size': MANIFEST_SIZE,
                    'name': 'manifest'
                })

                # Update manifest and add firmware regions
                for region in manifest["regions"]:
                    if "filename" not in region:
                        continue
                    if region.get("psram_dst") is not None:
                        assert region["spiflash_src"] is None
                        region["spiflash_src"] = firmware_base
                        print(f"manifest: region {region['filename']}: spiflash_src set to 0x{firmware_base:x}")
                        regions_to_check.append({
                            'addr': firmware_base,
                            'size': region["size"],
                            'name': region['filename'],
                        })
                        firmware_base += region["size"]
                        firmware_base = (firmware_base + 0xFFF) & ~0xFFF

                # Write updated manifest
                manifest_path = os.path.join(tmpdir, "manifest.json")
                print(f"\nFinal manifest contents:\n{json.dumps(manifest, indent=2)}")
                with open(manifest_path, "w") as f:
                    json.dump(manifest, f)

                # Collect all commands
                commands_to_run.append(
                    flash_file(
                        os.path.join(tmpdir, "top.bit"),
                        bitstream_addr,
                        dry_run=True
                    )
                )
                commands_to_run.append(
                    flash_file(
                        manifest_path,
                        manifest_addr,
                        "raw",
                        dry_run=True
                    )
                )
                for region in manifest["regions"]:
                    if "filename" not in region or "spiflash_src" not in region:
                        continue
                    commands_to_run.append(
                        flash_file(
                            os.path.join(tmpdir, region["filename"]),
                            region["spiflash_src"],
                            "raw",
                            dry_run=True
                        )
                    )

            # Print all regions
            print("\nRegions to be flashed:")
            for region in sorted(regions_to_check, key=lambda r: r['addr']):
                aligned_size = (region['size'] + FLASH_PAGE_SIZE - 1) & ~(FLASH_PAGE_SIZE - 1)
                print(f"  {region['name']}:")
                print(f"    start: 0x{region['addr']:x}")
                print(f"    end:   0x{region['addr']+aligned_size-1:x}")

            # Check for overlaps before proceeding
            has_overlap, error_msg = check_region_overlaps(regions_to_check, slot)
            if has_overlap:
                print(f"Error: {error_msg}")
                sys.exit(1)

            # Show all commands and get confirmation
            print("\nThe following commands will be executed:")
            for cmd in commands_to_run:
                print("\t$", " ".join(cmd))

            if not noconfirm:
                response = input("\nProceed with flashing? [y/N] ")
                if response.lower() != 'y':
                    print("Aborting.")
                    sys.exit(0)

            # Execute all commands
            print("\nExecuting flash commands...")
            for cmd in commands_to_run:
                subprocess.check_call(cmd)

            print("\nFlashing completed successfully")

def main():
    parser = argparse.ArgumentParser(description="Flash Tiliqua bitstream archives")
    parser.add_argument("archive", help="Path to bitstream archive (.tar.gz)")
    parser.add_argument("--slot", type=int, help="Slot number (0-7) for bootloader-managed bitstreams")
    parser.add_argument("--noconfirm", action="store_true", help="Do not ask for confirmation before flashing")

    args = parser.parse_args()

    if not os.path.exists(args.archive):
        print(f"Error: Archive not found: {args.archive}")
        sys.exit(1)

    if args.slot is not None and not 0 <= args.slot < MAX_SLOTS:
        print(f"Error: Slot must be between 0 and {MAX_SLOTS-1}")
        sys.exit(1)

    flash_archive(args.archive, args.slot, args.noconfirm)

if __name__ == "__main__":
    main() 

File: gateware/flash.js
import { runOpenFPGALoader, Exit } from 'https://cdn.jsdelivr.net/npm/@yowasp/openfpgaloader/gen/bundle.js';
import { Archive } from 'https://cdn.jsdelivr.net/npm/libarchive.js@2.0.2/+esm';

// Constants from the Python script
const BOOTLOADER_BITSTREAM_ADDR = 0x000000;
const SLOT_BITSTREAM_BASE = 0x100000;
const SLOT_SIZE = 0x100000;
const MANIFEST_SIZE = 1024;
const FIRMWARE_BASE_SLOT0 = 0x1C0000;
const FLASH_PAGE_SIZE = 1024;
const MAX_ERROR_LINES = 100;

// Store loaded archives
const loadedArchives = new Map();

// Add this to store message history for each slot
const messageHistory = new Map();

// Make handleFlash global
window.handleFlash = async function(slotId) {
    const slotData = loadedArchives.get(slotId.toString());
    if (!slotData) return;
    
    try {
        // Collect all commands first
        const commands = [];
        for (const region of slotData.regions) {
            const fileData = await slotData.archive.get(region.filename)?.arrayBuffer();
            if (fileData) {
                commands.push({
                    args: [
                        "-c", "dirtyJtag",
                        "-f",
                        "-o", `${region.addr.toString(16)}`,
                        "--file-type", region.filename.endsWith('.bit') ? "bit" : "raw"
                    ],
                    data: new Uint8Array(fileData),
                    name: region.filename,
                    addr: region.addr
                });
            }
        }

        // Show commands and get confirmation
        showError(slotId, "The following commands will be executed:\n" + 
            commands.map(cmd => 
                `openFPGALoader ${cmd.args.join(' ')} ${cmd.name}`
            ).join('\n')
        );

        /*
        if (!confirm("Proceed with flashing?")) {
            showError(slotId, "Flash cancelled");
            return;
        }
        */

        // Execute commands
        for (const cmd of commands) {
            try {
                // Create a virtual file for the data
                const filesIn = {
                    'data': cmd.data
                };
                
                // Add the filename to the args
                const args = [...cmd.args, 'data'];
                
                // Run openFPGALoader with the virtual file
                await runOpenFPGALoader(args, filesIn, {
                    stdout: (data) => {
                        if (data) {
                            const text = new TextDecoder().decode(data);
                            showError(slotId, text);
                        }
                    },
                    stderr: (data) => {
                        if (data) {
                            const text = new TextDecoder().decode(data);
                            showError(slotId, `stderr: ${text}`);
                        }
                    }
                });
            } catch (error) {
                if (error instanceof Exit) {
                    throw new Error(`Command failed with exit code ${error.code}`);
                }
                throw error;
            }
        }
        
        showError(slotId, "Flash completed successfully!");
    } catch (error) {
        showError(slotId, `Flash failed: ${error.message}`);
    }
};

// Initialize all event handlers after DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
    // File input handlers
    document.querySelectorAll('input[type="file"]').forEach(input => {
        input.addEventListener('change', async (e) => {
            const slotId = e.target.id.split('-')[1];
            const file = e.target.files[0];
            if (file) {
                try {
                    await loadArchive(file, slotId);
                    // Fix bootloader button selection
                    const button = slotId === 'bootloader' 
                        ? document.querySelector('#bootloader button')
                        : document.querySelector(`#slot-${slotId} button`);
                    if (button) {
                        button.disabled = false;
                    }
                } catch (error) {
                    showError(slotId, error.message);
                }
            }
        });
    });

    // Flash button handlers
    document.querySelectorAll('.flash-button').forEach(button => {
        button.addEventListener('click', () => {
            // Fix bootloader slot ID handling
            const slot = button.closest('.slot');
            const slotId = slot.id === 'bootloader' ? 'bootloader' : slot.id.replace('slot-', '');
            handleFlash(slotId);
        });
    });
});

// Add this helper class at the top of the file
class TarArchive {
    constructor() {
        this.files = new Map();
    }

    set(name, data) {
        this.files.set(name, data);
    }

    get(name) {
        const data = this.files.get(name);
        if (!data) return null;
        
        return {
            arrayBuffer: () => Promise.resolve(data),
            text: () => Promise.resolve(new TextDecoder().decode(data))
        };
    }
}

async function loadArchive(file, slotId) {
    const archive = await readTarGz(file);
    
    // Parse manifest
    const manifest = JSON.parse(await archive.get('manifest.json').text());
    
    // Collect regions
    const regions = [];
    const isBootloader = slotId === 'bootloader';
    const slot = isBootloader ? null : parseInt(slotId);
    
    if (isBootloader) {
        // XIP firmware handling
        regions.push({
            name: 'bootloader bitstream',
            filename: 'top.bit',
            addr: BOOTLOADER_BITSTREAM_ADDR,
            size: (await archive.get('top.bit').arrayBuffer()).byteLength,
        });
        
        // Add XIP firmware regions
        for (const region of manifest.regions || []) {
            if (region.filename) {
                regions.push({
                    name: `firmware '${region.filename}'`,
                    filename: region.filename,
                    addr: region.spiflash_src,
                    size: region.size,
                });
            }
        }
    } else {
        // Regular slot handling
        const slotBase = SLOT_BITSTREAM_BASE + (slot * SLOT_SIZE);
        const firmwareBase = FIRMWARE_BASE_SLOT0 + (slot * SLOT_SIZE);
        
        regions.push({
            name: 'bitstream',
            filename: 'top.bit',
            addr: slotBase,
            size: (await archive.get('top.bit').arrayBuffer()).byteLength,
        });
        
        regions.push({
            name: 'manifest',
            filename: 'manifest.json',
            addr: (slotBase + SLOT_SIZE) - MANIFEST_SIZE,
            size: MANIFEST_SIZE,
        });
        
        let currentFirmwareBase = firmwareBase;
        for (const region of manifest.regions || []) {
            if (region.filename && region.psram_dst !== undefined) {
                regions.push({
                    name: region.filename,
                    filename: region.filename,
                    addr: currentFirmwareBase,
                    size: region.size,
                });
                currentFirmwareBase += region.size;
                currentFirmwareBase = (currentFirmwareBase + 0xFFF) & ~0xFFF;
            }
        }
    }
    
    // Check for overlaps
    const hasOverlap = checkRegionOverlaps(regions, slot);
    if (hasOverlap[0]) {
        throw new Error(hasOverlap[1]);
    }
    
    // Store archive and regions
    loadedArchives.set(slotId, { archive, regions });
    
    // Display regions
    displayRegions(slotId, regions);
}

function checkRegionOverlaps(regions, slot) {
    const alignedRegions = [];
    
    for (const r of regions) {
        const start = r.addr;
        const size = (r.size + FLASH_PAGE_SIZE - 1) & ~(FLASH_PAGE_SIZE - 1);
        const end = start + size;
        alignedRegions.push([start, end, r.name]);
        
        if (slot !== null) {
            const slotStart = Math.floor(start / SLOT_SIZE) * SLOT_SIZE;
            const slotEnd = slotStart + SLOT_SIZE;
            if (end > slotEnd) {
                return [true, `Region '${r.name}' exceeds slot boundary: ends at 0x${end.toString(16)}, slot ends at 0x${slotEnd.toString(16)}`];
            }
        }
    }
    
    alignedRegions.sort((a, b) => a[0] - b[0]);
    
    for (let i = 0; i < alignedRegions.length - 1; i++) {
        const currEnd = alignedRegions[i][1];
        const nextStart = alignedRegions[i + 1][0];
        if (currEnd > nextStart) {
            return [true, `Overlap detected between '${alignedRegions[i][2]}' (ends at 0x${currEnd.toString(16)}) and '${alignedRegions[i + 1][2]}' (starts at 0x${nextStart.toString(16)})`];
        }
    }
    
    return [false, ""];
}

function displayRegions(slotId, regions) {
    const regionsDiv = document.getElementById(`regions-${slotId}`);
    regionsDiv.innerHTML = regions.map(r => {
        const alignedSize = (r.size + FLASH_PAGE_SIZE - 1) & ~(FLASH_PAGE_SIZE - 1);
        return `${r.name}:<br>` +
               `&nbsp;&nbsp;start: 0x${r.addr.toString(16)}<br>` +
               `&nbsp;&nbsp;end: 0x${(r.addr + alignedSize - 1).toString(16)}`;
    }).join('<br><br>');
}

function showError(slotId, message) {
    // Get or create message history array for this slot
    if (!messageHistory.has(slotId)) {
        messageHistory.set(slotId, []);
    }
    const history = messageHistory.get(slotId);
    
    // Split new message into lines and add each line to history
    const newLines = message.split('\n');
    history.push(...newLines);
    
    // Keep only the last MAX_ERROR_LINES lines
    while (history.length > MAX_ERROR_LINES) {
        history.shift();
    }
    
    // Update the display
    const errorDiv = document.getElementById(`error-${slotId}`);
    errorDiv.innerHTML = history.map(line => 
        `<div class="message-line">${line}</div>`
    ).join('');
    
    // Scroll to bottom
    errorDiv.scrollTop = errorDiv.scrollHeight;
}

async function readTarGz(file) {
    // Create archive object for our use
    const archive = new TarArchive();
    
    // Read the file
    const fileData = await file.arrayBuffer();
    
    // Decompress gzip
    const inflated = pako.inflate(new Uint8Array(fileData));
    
    // Parse tar - process 512 byte blocks
    let offset = 0;
    while (offset < inflated.length) {
        // Read header block
        const header = inflated.slice(offset, offset + 512);
        
        // Check for end of archive (empty block)
        if (header.every(byte => byte === 0)) {
            break;
        }
        
        // Parse filename (100 bytes)
        const filename = new TextDecoder().decode(header.slice(0, 100)).split('\0')[0];
        
        // Parse file size (12 bytes, octal string)
        const sizeStr = new TextDecoder().decode(header.slice(124, 136)).trim();
        const size = parseInt(sizeStr, 8);
        
        // Move past header
        offset += 512;
        
        // Read file content
        if (size > 0) {
            const content = inflated.slice(offset, offset + size);
            archive.set(filename, content);
            
            // Move to next 512-byte boundary
            offset += (Math.floor((size + 511) / 512) * 512);
        }
    }
    
    return archive;
} 